itests/kit/node_unmanaged.go

```diff
@@ -93,271 +132,465 @@ func NewTestUnmanagedMiner(t *testing.T, full *TestFullNode, actorAddr address.A
-	head, err := tm.FullNode.ChainHead(ctx)
-	require.NoError(tm.t, err)
+type OnboardOpt func(opts *onboardOpt) error
+type onboardOpt struct {
+	modifyNIActivationsBeforeSubmit func([]miner14.SectorNIActivationInfo) []miner14.SectorNIActivationInfo
+	requireActivationSuccess        bool
+}
 
-	p, err := tm.FullNode.StateMinerPower(ctx, tm.ActorAddr, head.Key())
-	require.NoError(tm.t, err)
+func WithModifyNIActivationsBeforeSubmit(f func([]miner14.SectorNIActivationInfo) []miner14.SectorNIActivationInfo) OnboardOpt {
+	return func(opts *onboardOpt) error {
+		opts.modifyNIActivationsBeforeSubmit = f
+		return nil
+	}
+}
 
-	return p
+func WithRequireActivationSuccess(requireActivationSuccess bool) OnboardOpt {
+	return func(opts *onboardOpt) error {
+		opts.requireActivationSuccess = requireActivationSuccess
+		return nil
+	}
 }
 
-func (tm *TestUnmanagedMiner) AssertPower(ctx context.Context, raw uint64, qa uint64) {
+// OnboardSectors onboards the specified number of sectors to the miner using the specified proof
+// type. If `withPieces` is true and the proof type supports it, the sectors will be onboarded with
+// pieces, otherwise they will be CC.
+//
+// This method is synchronous but each sector is prepared in a separate goroutine to speed up the
+// process with real proofs.
+func (tm *TestUnmanagedMiner) OnboardSectors(
+	proofType abi.RegisteredSealProof,
+	withPieces bool,
+	count int,
+	opts ...OnboardOpt,
+) []OnboardedSector {
+
 	req := require.New(tm.t)
-	p := tm.CurrentPower(ctx)
-	tm.t.Logf("Miner %s RBP: %v, QaP: %v", p.MinerPower.QualityAdjPower.String(), tm.ActorAddr, p.MinerPower.RawBytePower.String())
-	req.Equal(raw, p.MinerPower.RawBytePower.Uint64())
-	req.Equal(qa, p.MinerPower.QualityAdjPower.Uint64())
+
+	options := onboardOpt{}
+	for _, o := range opts {
+		req.NoError(o(&options))
+	}
+
+	sectors := make([]sectorInfo, count)
+	var wg sync.WaitGroup
+
+	// Wait for the seal randomness to be available (we can only draw seal randomness from
+	// tipsets that have already achieved finality)
+	sealRandEpoch, err := tm.waitPreCommitSealRandomness(proofType)
+	req.NoError(err)
+
+	// For each sector, run PC1, PC2, C1 an C2, preparing for ProveCommit. If the proof needs it we
+	// will also submit a precommit for the sector.
+	for i := 0; i < count; i++ {
+		wg.Add(1)
+		// We hold on to `sector`, adding new properties to it as we go along until we're finished with
+		// this phase, then add it to `sectors`
+		sector := tm.nextSector(proofType)
+		sector.sealRandomnessEpoch = sealRandEpoch
+
+		go func(idx int) {
+			assert := assert.New(tm.t)
+
+			if withPieces {
+				// Build a sector with non-zero pieces to onboard
+				if tm.mockProofs {
+					sector.pieces = []abi.PieceInfo{{
+						Size:     abi.PaddedPieceSize(tm.options.sectorSize),
+						PieceCID: fixedPieceCid,
+					}}
+				} else {
+					var err error
+					sector, err = tm.mkAndSavePiecesToOnboard(sector)
+					req.NoError(err)
```

Comment:
Replaced with the `assert` dance; thanks.

Yeah, so `require` is going to make a fatal and panic, with no opportunity for cleanup, `assert` will just fail and then let you deal with it later. So I've taken the generally recommended approach in this file of anywhere that is called from a goroutine other than the primary test one should either use `assert` or return the error so it can be dealt with properly.